hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java

```diff
@@ -0,0 +1,171 @@
+package com.hazelcast.spi.impl;
+
+import com.hazelcast.instance.Node;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.util.ConcurrencyUtil;
+import com.hazelcast.util.ConstructorFunction;
+
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Monitors the operation threads to find and collect data of long running operations.
+ */
+final class SlowOperationDetector extends Thread {
+
+    private static final long ONE_SECOND_IN_NANOS = TimeUnit.SECONDS.toNanos(1);
+
+    private final BasicOperationScheduler.OperationThread[] genericOperationThreads;
+    private final BasicOperationScheduler.OperationThread[] partitionOperationThreads;
+
+    private final CurrentOperationData[] genericCurrentOperationData;
+    private final CurrentOperationData[] partitionCurrentOperationData;
+
+    private final long slowOperationThresholdNanos;
+    private final long logPurgeIntervalNanos;
+    private final long logLifetimeNanos;
+
+    private final ILogger logger;
+
+    private final ConcurrentHashMap<Integer, SlowOperationLog> slowOperationLogs = new ConcurrentHashMap<Integer, SlowOperationLog>();
+
+    private long nowNanos;
+    private long nowMillis;
+
+    private volatile boolean running = true;
+
+    public SlowOperationDetector(Node node, BasicOperationScheduler scheduler) {
+        super(SlowOperationDetector.class.getSimpleName());
+        setDaemon(true);
+
+        genericOperationThreads = scheduler.genericOperationThreads;
+        partitionOperationThreads = scheduler.partitionOperationThreads;
+
+        genericCurrentOperationData = new CurrentOperationData[genericOperationThreads.length];
```

Comment:
I would do it like this:

genericCurrentOperationData = initGenericCurrentOperationData()

And like the init method take care of creating and filling the array. Now it is a bit smeared and makes the constructor longer than needs to be. The same things goes for the partition...array.